### PR TITLE
Add support for ROCm 5.5 and 5.6

### PR DESCRIPTION
**Pull request by Rombur**
This PR does 3 things:
 - starting with ROCm 5.5, we don't need to link with `clang_rt`. It's not obvious to me how to get the ROCm version inside CMake, just add a comment for now
 - Newer `amdclang` and Cray `clang` are based on Clang 16 which uses C++17 by default. However, HIP is considered a different language with a different default standard. Add a message if a user does not set `CMAKE_CXX_STANDARD=17` similar to what we used to do with C++14. Note that `hipcc` does not have this issue.
 - Disable some tests for ROCm 5.5 and 5.6 due to a compiler bug. AMD has already fixed the bug upstream.

### CODE REVIEW

**Review by crtrott — APPROVED** [review]

File: cmake/kokkos_arch.cmake [hunk 1]
```diff
@@ -210,6 +210,9 @@ GLOBAL_SET(KOKKOS_AMDGPU_OPTIONS)
 IF(KOKKOS_ENABLE_HIP)
   SET(AMDGPU_ARCH_FLAG "--offload-arch")
   IF(NOT KOKKOS_CXX_COMPILER_ID STREQUAL HIPCC)
+    IF (NOT CMAKE_CXX_STANDARD)
+      MESSAGE(FATAL_ERROR "Set CMAKE_CXX_STANDARD to 17")
```
Comment (masterleinad) [inline cmake/kokkos_arch.cmake:214]:
```suggestion
      MESSAGE(FATAL_ERROR "Kokkos requires CMAKE_CXX_STANDARD to be set to 17 or higher!")
```

File: core/unit_test/TestMathematicalSpecialFunctions.hpp [hunk 2]
```diff
@@ -1196,6 +1196,11 @@ struct TestComplexBesselI0K0Function {
     h_ref_cbk0(24) =
         Kokkos::complex<double>(1.413897840559108e-27, -1.851678917759592e+25);
 
+    // FIXME_HIP Disable the test when using ROCm 5.5 and 5.6 due to a known
+    // compiler bug
+#if (HIP_VERSION_MAJOR != 5) ||  \
+    ((HIP_VERSION_MAJOR == 5) && \
+     !((HIP_VERSION_MINOR == 5) || (HIP_VERSION_MINOR == 6)))
```
Comment (masterleinad) [inline core/unit_test/TestMathematicalSpecialFunctions.hpp:1203]:
```suggestion
#if !defined(KOKKOS_ENABLE_HIP) || (HIP_VERSION_MAJOR != 5) ||  \
     !((HIP_VERSION_MINOR == 5) || (HIP_VERSION_MINOR == 6))
```

File: core/unit_test/TestMathematicalSpecialFunctions.hpp [hunk 3]
```diff
@@ -1686,6 +1692,11 @@ struct TestComplexBesselH1Function {
     h_ref_ch11(24) =
         Kokkos::complex<double>(-5.430453818237824e-02, -1.530182458039000e-02);
 
+    // FIXME_HIP Disable the test when using ROCm 5.5 and 5.6 due to a known
+    // compiler bug
+#if (HIP_VERSION_MAJOR != 5) ||  \
+    ((HIP_VERSION_MAJOR == 5) && \
+     !((HIP_VERSION_MINOR == 5) || (HIP_VERSION_MINOR == 6)))
```
Comment (masterleinad) [inline core/unit_test/TestMathematicalSpecialFunctions.hpp:1699]:
```suggestion
#if !defined(KOKKOS_ENABLE_HIP) || (HIP_VERSION_MAJOR != 5) ||  \
     !((HIP_VERSION_MINOR == 5) || (HIP_VERSION_MINOR == 6))
```

File: core/unit_test/TestMathematicalSpecialFunctions.hpp [hunk 4]
```diff
@@ -1868,6 +1880,11 @@ struct TestComplexBesselH2Function {
     h_ref_ch21(24) =
         Kokkos::complex<double>(1.629136145471347e-01, +1.530182458039000e-02);
 
+    // FIXME_HIP Disable the test when using ROCm 5.5 and 5.6 due to a known
+    // compiler bug
+#if (HIP_VERSION_MAJOR != 5) ||  \
+    ((HIP_VERSION_MAJOR == 5) && \
+     !((HIP_VERSION_MINOR == 5) || (HIP_VERSION_MINOR == 6)))
```
Comment (masterleinad) [inline core/unit_test/TestMathematicalSpecialFunctions.hpp:1887]:
```suggestion
#if !defined(KOKKOS_ENABLE_HIP) || (HIP_VERSION_MAJOR != 5) ||  \
     !((HIP_VERSION_MINOR == 5) || (HIP_VERSION_MINOR == 6))
```

**Review by masterleinad — APPROVED** [review]

**Comment by crtrott** [comment]
Timed out on HIP? Don't see how or why but gonna rerun now.

**Comment by crtrott** [comment]
Retest this please.

**Comment by Rombur** [comment]
> Timed out on HIP? Don't see how or why but gonna rerun now.

The two HIP builds passed

**Comment by cwsmith** [comment]
:+1: I hit the second point/issue with 4.1.00 on Frontier with rocm 5.5.1 and amdclang (need to pass  `-DCMAKE_CXX_STANDARD=17`).

**Pull request merged**